smtp/CMakeLists.txt

```diff
@@ -4,7 +4,24 @@ endif(APPLE)
 
 # Add C library
 add_library(lib SHARED lib.c smtpc.c)
-target_link_libraries(lib curl ${MSGPUCK_LIBRARIES})
+
+set(TARANTOOL_CURL_PATH "/tarantool/curl")
+find_path(HAS_BUNDLED_LIBCURL NAMES ${TARANTOOL_CURL_PATH})
+
+if(USE_TARANTOOL_CURL)
+  if (NOT HAS_BUNDLED_LIBCURL)
+    message(WARNING "Your Tarantool version hasn't bundled libCURL. Use system \
+            libCURL.")
+    target_link_libraries(lib curl ${MSGPUCK_LIBRARIES})
+  else()
+    message("-- Use Tarantool bundled libCURL.")
+    target_include_directories(lib PRIVATE
+                               ${HAS_BUNDLED_LIBCURL}/${TARANTOOL_CURL_PATH})
+  endif()
+else()
+  target_link_libraries(lib curl ${MSGPUCK_LIBRARIES})
+endif()
```

Comment:
Let's start from usage scenarious:

1. A developer of the module can either install libcurl or use bundled one. Any behaviour is okay.
2. An application developer can install the module using luarocks / tarantoolctl rocks. The module will be used with currently installed tarantool. It is not redistributable to other systems. However updating of tarantool to a newer version should not break the module.
3. Building of a package. It seems, it will be okay to always link against the bundled libcurl. I guess such build will work as for tarantool with bundled libcurl as well as for one with the system one (except, maybe, cases with bundled libcurl without smtp(s) support).

So we need the following ways to build:

* Default: prefer libcurl bundled into tarantool, but fallback to a system libcurl.
* For packages: always use bundled libcurl; bump tarantool and tarantool-devel/dev dependecies.

So packages should pass an option like `-DTARANTOOL_CURL_REQUIRED=ON` to cmake (and the option should be disabled by default).

Next, I don't see your testing plan and results of the testing. Please, elaborate. Otherwise we'll duplicate effort.

I would not duplicate `target_link_libraries()` calls, but rather set `${CURL_LIBRARIES}` and `${CURL_INCLUDE_DIRS}`. Your way is error-prone. For example, one of branches missed `target_link_libraries()` at all, so the dependency on msgpuck is not set.

Hm... It seems, msgpuck is not used in the module, so, please, remove it first (in a separate commit) from `debian/control` and `smtp/CMakeLists.txt`. That's confusing.

I would write it like so (I didn't test):

```cmake
find_path(TARANTOOL_CURL_INCLUDE_DIR
    NAMES curl/curl.h
    PATH_SUFFIXES tarantool)
if (TARANTOOL_CURL_INCLUDE_DIR)
    message(STATUS "Link against tarantool's bundled libcurl.")
    set(CURL_LIBRARIES)
    set(CURL_INCLUDE_DIRS "${TARANTOOL_CURL_INCLUDE_DIR}")
elseif (TARANTOOL_CURL_REQUIRED)
    # XXX: Mention a tarantool versions necessary to succussfully
    # build the module.
    message(FATAL_ERROR "<...>")
else()
    message(WARNING "Installed tarantool version does not ship libcurl headers. Fallback to linking with system libcurl.")
    set(CURL_LIBRARIES curl)
    set(CURL_INCLUDE_DIRS)
endif()
target_include_directories(lib PRIVATE ${CURL_INCLUDE_DIRS})
target_link_libraries(lib PRIVATE ${CURL_LIBRARIES})
```

Reply:
I tested current branch with [2.8](https://github.com/tarantool/tarantool/tree/romanhabibov/curl-headers) and [1.10](https://github.com/tarantool/tarantool/tree/romanhabibov/curl-smtp_1_10):
developer's build works fine.

Reply:
It seems, we unable to support all 2.5-2.7 releases without a **runtime** logic, whether to use tarantool's built-in libcurl or system's one. I'll share details below.

Regarding this particular commit:

1. I would remove the `TARANTOOL_CURL_REQUIRED=ON` option from the rockspec. Packages are prebuilt, but installation using luarocks / tarantoolctl rocks assumes the build step. So the additional requirement for presence of the bundled libcurl in tarantool would forbid installation of the module on all existing ('old') tarantool versions.
2. I don't see a usage scenario for `-DUSE_TARANTOOL_CURL=OFF -DTARANTOOL_CURL_REQUIRED=OFF`, that's why I proposed the sole `-DTARANTOOL_CURL_REQUIRED=ON/OFF`. It seems, it was not obvious from my wording, sorry. (And the options are not orthogonal this way: `-DUSE_TARANTOOL_CURL=OFF -DTARANTOOL_CURL_REQUIRED=ON` looks useless and is not handled gracefully now). I would drop the `USE_TARANTOOL_CURL` option.
3. lubCURL/libCURL -> libcurl (see the explanation [here](https://lists.tarantool.org/pipermail/tarantool-patches/2021-March/023177.html)).
4. Without `NO_DEFAULT_PATH` the `find_path()` call may return a system's libcurl header, so we'll never fall into the `else()` branch.

Anyway, a build-time check does not look sufficient, because many tarantool releases in the 2.5-2.7 series expose libcurl functions without smtp(s) support. A function from the executable have precedence over one from system's libcurl even if we list /usr/lib/libcurl.so in the NEEDED field of the .so/dylib file.